File: Python/bir işlem/business/gameManager/iGameManager.py
from abc import ABC, abstractclassmethod as interface
from typing import Literal

from entity.game_data import GameData
from entity.player import Player


class IGameManager(ABC):
    """GameManager sınıfı için gerekli olan metotların imzalarını tutar.
    
    @category: Interface, Manager"""

    @property  # type: ignore
    @interface
    def game_data(self) -> GameData:
        raise NotImplementedError(
            "Implement 'game_data' method from IGameManager")

    @property  # type: ignore
    @interface
    def player(self) -> Player:
        raise NotImplementedError(
            "Implement 'player' method from IGameManager")

    @interface
    def create_numbers(self) -> None:
        raise NotImplementedError(
            "Implement 'create_numbers' method from IGameManager")

    @interface
    def choice_npc(self) -> None:
        raise NotImplementedError(
            "Implement 'choice_npc' method from IGameManager")

    @interface
    def isFinished(self, value: int) -> bool:
        raise NotImplementedError(
            "Implement 'isFinished' method from IGameManager")

    @interface
    def process(self, gelen_sonuc: float, gelen_sayi: float, gelen_cikti: str) -> tuple[float, str]:
        raise NotImplementedError(
            "Implement 'process' method from IGameManager")

    @interface
    def information_message(self) -> tuple[bool, int]:
        raise NotImplementedError(
            "Implement 'information_message' method from IGameManager")

    @interface
    def calculate(self, tekrar_sayisi: int) -> Literal[True] | None:
        raise NotImplementedError(
            "Implement 'calculate' method from IGameManager")

File: Java/bir işlem/business/IGameManager.java
package business;

import entity.GameData;
import entity.Player;

/**
 * Oyun iş katmanı için kullanılacak operasyonların imzasını tutar
 * 
 * @category Business, Interface
 */
public interface IGameManager {
    /**
     * Oyun verilerinin tutulacağı varlık sınıfını verir.
     * 
     * @return gameData -> Oyun Verileri (GameData)
     * @see GameData
     */
    GameData gameData();

    /**
     * NPC (Non-Player Character) varlık sınıfını verir.
     * 
     * @return player -> NPC (Player)
     * @see Player
     */
    Player player();

    /**
     * Oyun içerisinde kullanılacak olan sayıların türetilmesini sağlar.
     * 
     * @apiNote 5 tane 1 basamaklı
     * @apiNote 1 tane 2 basamaklı
     * @apiNote 1 tane 3 basamaklı
     */
    void createNumbers();

    /**
     * Oyunu oynayacak NPC (Non-Player Character)'i türetir.
     */
    void choiceNpc();

    /**
     * Yapılan işlemler sonucunda ulaşılan sonucun, ulaşılmak istenen sonuç
     * aralığında olup olmadığını kontrol eder.
     * 
     * @param bulunanSonuc -> Hesaplamalar sonucu ulaşılan sonuç (float)
     * @return boolean -> bulunanSonuc istenilen aralıktaysa True, değilse False
     */
    boolean isFinished(float bulunanSonuc);

    /**
     * Tüm hesaplamaları rasgele olarak yapmaya yarar.
     * 
     * @param gelenSonuc -> Bir önceki adımda bulunmuş olan sonuç değeri (float)
     * @param gelenSayi  -> işleme tabi tutulacak sayı değeri (float)
     * @param gelenCikti -> daha önceki işlemlerin geçmişi (String)
     */
    void process(float gelenSonuc, float gelenSayi, String gelenCikti);

    /**
     * Rasgele üretilmiş sayıları, yine rasgele olarak işlemlere tabi tutup
     * istenilen sayıya ulaşmaya çalışır.
     * 
     * @param tekrarSayisi -> İşlemlerin kaç defa tekrarlanacağını (denemenin bir
     *                     seferde kaç defa yapılacağını) belirtir. (int)
     */
    void calculate(int tekrarSayisi);

}
